src/storage/data_table.cpp

```diff
@@ -68,11 +69,18 @@ DataTable::SlotIterator &DataTable::SlotIterator::operator++() {
   common::SpinLatch::ScopedSpinLatch guard(&table_->blocks_latch_);
   // Jump to the next block if already the last slot in the block.
   if (current_slot_.GetOffset() == table_->accessor_.GetBlockLayout().NumSlots() - 1) {
-    ++block_;
-    // Cannot dereference if the next block is end(), so just use nullptr to denote
-    current_slot_ = {block_ == table_->blocks_.end() ? nullptr : *block_, 0};
+    if (num_advances_ > 0 || num_advances_ == SlotIterator::TILL_END) {
+      // Advance to the next block.
+      ++block_index_;
+      num_advances_ = num_advances_ == SlotIterator::TILL_END ? num_advances_ : num_advances_ - 1;
+      // Cannot dereference if the next block is end(), so just use nullptr to denote
+      current_slot_ = {block_index_ >= table_->blocks_.size() ? nullptr : table_->blocks_[block_index_], 0};
+    } else {
+      // Done advancing, time to give up.
+      current_slot_ = {nullptr, 0};
+    }
   } else {
-    current_slot_ = {*block_, current_slot_.GetOffset() + 1};
+    current_slot_ = {table_->blocks_[block_index_], current_slot_.GetOffset() + 1};
```

Comment:
The "block" for this tuple slot is not changing so there is no reason to hit the vector and take a possible cache miss inside a critical section.  It would be better to use `current_slot_.GetBlock()` which is pretty much guaranteed to be a cache hit (if not already in a register).